paddle/framework/operator.h

```diff
@@ -347,6 +341,154 @@ template <>
 std::vector<Tensor*> ExecutionContext::MultiOutput<Tensor>(
     const std::string& name) const;
 
+class BlockDesc {
+ public:
+  explicit BlockDesc(const std::map<std::string, VarDesc*>& var_descs)
+      : var_descs_(var_descs) {}
+  ~BlockDesc() {}
+
+  VarDesc* GetVar(const std::string& name) const {
+    PADDLE_ENFORCE(var_descs_.count(name) == 1, "%s must be in Block", name);
+    return var_descs_.at(name);
+  }
+
+  bool HasVar(const std::string& name) const {
+    return var_descs_.count(name) > 0;
+  }
+
+ private:
+  const std::map<std::string, VarDesc*>& var_descs_;
+};
+
+class CompileTimeInferShapeContext : public InferShapeContextBase {
+ public:
+  CompileTimeInferShapeContext(const OperatorBase& op,
+                               const BlockDesc& block_desc)
+      : op_(op), block_desc_(block_desc) {}
+
+  bool HasInput(const std::string& name) const {
+    return block_desc_.HasVar(name);
+  }
+
+  bool HasOutput(const std::string& name) const {
+    return block_desc_.HasVar(name);
+  }
+
+  DDim GetInputDim(const std::string& name) const {
+    return GetDim(op_.Input(name));
+  }
+
+  void SetInputDim(const std::string& name, const DDim& dim) const {
+    SetDim(op_.Input(name), dim);
+  }
+
+  DDim GetOutputDim(const std::string& name) const {
+    return GetDim(op_.Output(name));
+  }
+
+  void SetOutputDim(const std::string& name, const DDim& dim) const {
```

Comment:
Set methods should change private data members, so it cannot be `const`

Reply:
done